Docs/documentation 2.0 (#158)
sse-secure-systems/connaisseur: docs/features/alerting.md

# Alerting

Connaisseur can send notifications on admission decisions to basically every REST
endpoint that accepts JSON payloads.

## Supported Interfaces

> Slack, Opsgenie and Keybase have preconfigured payloads that are ready to use.

Reviewer: ```suggestion
Slack, Opsgenie and Keybase have pre-configured payloads that are ready to use.
```

| `alerting.<category>.receiver_url`                 | string                                                | -                 | :heavy_check_mark: | url of alert-receiving endpoint                                                                    |
| `alerting.<category>.priority`                     | int                                                   | `3`               |                    | priority of alert (to enable fitting Connaisseur alerts into alerts from other sources)            |
| `alerting.<category>.custom_headers`               | list[string]                                          | -                 |                    | additional headers required by alert-receiving endpoint                                            |
| `alerting.<category>.payload_fields`               | subyaml                                               | -                 |                    | enables specifying additional (`yaml`) key-value pairs to be appended to alert payload (as `json`) |

> | `alerting.<category>.fail_if_alert_sending_fails`  | bool                                                  | `False`           |                    | `True` will make Connaisseur deny images if the corresponding alert cannot be successfully sent    |

Reviewer: Other config bools are false/true instead of False/True. I guess this is due to how it is handled under the hood. Would that be something to change at some point? Out of scope for this PR, but if you think reasonable, maybe new issue

Author: feel free to create an issue

<sup>*basename of the custom template file in `helm/alerting_payload_templates` without file extension </sup>

_Notes_:

- The value for `template` needs to match an existing file of the pattern
`helm/alert_payload_templates/<template>.json`; so if you want to use a predefined
one it needs to be one of `slack`, `keybase` or `opsgenie`.
- For Opsgenie you need to configure an additional
  `["Authorization: GenieKey <Your-Genie-Key>"]` header.
- `fail_if_alert_sending_fails` only comes into play for requests that Connaisseur would have admitted as other requests would have been denied in the first place. The setting can come handy if you want to run Connaisseur in detection mode but still make sure that you get notified about what is going on in your cluster. **However, this setting will significantly impact cluster interaction for everyone (i.e. block any cluster change associated to an image) if the alert sending fails permanently, e.g. accidental deletion of your Slack Webhook App, GenieKey expired...**

## Example

> For example, if you would like to receive notifications in Keybase whenever Connaisseur admits a request to your cluster, your alerting configuration would look similar to to following snippet:

Reviewer: ```suggestion
For example, if you would like to receive notifications in Keybase whenever Connaisseur admits a request to your cluster, your alerting configuration would look similar to the following snippet:
```

```
alerting:
  admit_request:
    templates:
      - template: keybase
        receiver_url: https://bots.keybase.io/webhookbot/<Your-Keybase-Hook-Token>
```

## Additional notes

### Creating a custom template

Along the lines of the templates that already exist you can easily define
custom templates for other endpoints. The following variables can be rendered
during runtime into the payload:

- `alert_message`
- `priority`
- `connaisseur_pod_id`
- `cluster`
- `timestamp`
- `request_id`
- `images`

Referring to any of these variables in the templates works by Jinja2 notation
(e.g. `{{ timestamp }}`). You can update your payload dynamically by adding payload

> fields in `yaml` presentation in the `payload_fields` key which will be translated

Reviewer: ```suggestion
fields in `yaml` representation in the `payload_fields` key which will be translated
```
maybe this?

> to JSON by helm as is. If your REST endpoint requires particular headers, your can

Reviewer: ```suggestion
to JSON by Helm as is. If your REST endpoint requires particular headers, you can
```
